### DISCUSSION TITLE
Persisting state with actors

### DISCUSSION THREAD

**Discussion by DirkWolthuis** [statelyai/xstate#1515] [Q&A] [answered]
Hi all, 

What is the recommended way to persists the state of a machine that uses actors? I've made a [codesandbox](https://codesandbox.io/s/laughing-tdd-disyp?file=/src/index.js:0-2506) that demonstrates my problem. When saving the state to localstorage, the actors get reduced to an object with only a id key.

**Comment by davidkpiano** [comment, marked as answer]
If you want services/actors to be restarted, the best thing to do would be to try an **event sourcing approach** instead: https://codesandbox.io/s/event-sourcing-example-xe952?file=/src/index.js

```js
// Example: you can use localStorage instead
const persistedEvents = localStorage.getItem('events');

const service = interpret(machine)
  .onTransition(state => {
    if (state.changed !== undefined) {
      // push event to local storage
    }
  });

service.start();

// replay events
persistedEvents.forEach(event => {
  service.send(event);
});
```

**Reply by DirkWolthuis** [reply]
Even if i'm just interested in the last state of the actors? If I understand correctly, if the actor does an API call to get some data in its context, with event sourcing it will make that call as the events do a replay?

**Reply by davidkpiano** [reply]
This is where it gets tricky. We're still thinking about the best ways to do this in v5.

For prior art, see how Akka does it: https://doc.akka.io/docs/akka/current/typed/persistence.html

**Reply by Koleok** [reply]
[nact](https://github.com/nactio/nact) also has a persistence approach that includes snapshots and a [postgres adapter](https://github.com/ncthbrt/nact-persistence-postgres) that might be worth drawing inspiration from

**Reply by nirname** [reply]
What if I am using after and other delayed features. Should I wait for all the events to be replayed with all the pauses?

**Comment by javiermadueno** [comment]
When replying the events, what happens to those states that instantiate a service that makes a POST request? Is it executed again or it's cancelled because it's immediately recives the `done.invoke` event? Can we be sure the POST request is not going to be made?
